# Exercise Informed Search Algorithms #


In the last session, we implemented different systematic search strategies. If we want to find paths between different cities in a map, we can use additional information to guide our search. We don't rely on the 'blind' search and can implement more efficient algorithms, that consider the coordinates of the cities for example.

Implement the following algorithms and answer the same questions as you did for the systematic search algorithms.

1. A* Algorithm
1. IDA* Search

In [28]:
from sbb import SBB
from heapq import heappush, heappop
from search_solution import *


In [29]:
import os

# just necessary for vscode to be able to debug... 
os.chdir('/Users/jabbathegut/PipenvPythonProjects/AISO/07_HeuristicAndLocalSearchAlgorithms')

In [30]:
def a_star_algorithm(problem, sbb):
    # define empty frontier
    frontier = []
    # keep track of visited states
    visited = []
    # add initial node to frontier with piority 0
    node = Node(problem.initial)
    heappush(frontier, (0, node))
    
    if problem.goal_test(node.state):
        return node
    
    while True:
        node = heappop(frontier)[1]
        
        for action in problem.actions(node.state):
            child_node = node.child_node(problem, action)
            heuristics = (sbb.get_distance_between(child_node.state, problem.goal)) + child_node.path_cost
            
            if problem.goal_test(child_node.state):
                print('Reached goal destination: ' + child_node.state)
                print('Nodes visited: ' + str(len(visited)))
                print('Max nodes stored: ' + str(len(visited)) + str(len(frontier)))
                return child_node
            if not child_node.state in visited:
                heappush(frontier, (heuristics, child_node))
                visited.append(child_node.state)

In [31]:
def ida_star_algorithm(problem, sbb):
    boundary = get_heuristic(Node(problem.initial), problem, sbb)
    
    return search(problem, sbb, boundary)
    

def search(problem, sbb, boundary):
    # define empty frontier
    frontier = []
    # keep track of visited states
    visited = []
    # add initial node to frontier with piority 0
    node = Node(problem.initial)
    heappush(frontier, (0, node))
    
    if problem.goal_test(node.state):
        return node
    
    # check if first node is in range of the cost boundary 
    heuristics = get_heuristic(node, problem, sbb)
    
    # check costs
    while boundary >=  heuristics:
        # get all child nodes, calculate costs
        for action in problem.actions(node.state):
            child_node = node.child_node(problem, action)
            heuristics = get_heuristic(child_node, problem, sbb)
            
            if problem.goal_test(child_node.state):
                print('Reached goal destination: ' + child_node.state)
                print('Nodes visited: ' + str(len(visited)))
                print('Max nodes stored: ' + str(len(visited)) + str(len(frontier)))
                return child_node
            if not child_node.state in visited:
                heappush(frontier, (heuristics, child_node))
                visited.append(child_node.state)
                
        # get node with lowest cost and calculate new costs
        node = heappop(frontier)[1]
        heuristics = get_heuristic(node, problem, sbb)
    
    # this is returned when search is cancelled, heuristics is the new cost boundary for next iteration
    return search(problem, sbb, heuristics)

def get_heuristic(node, problem, sbb):
    return sbb.get_distance_between(node.state, problem.goal) + node.path_cost

In [32]:
def print_info_about_search(goal_node):
    node = goal_node
    if node:
        print("The search algorithm reached " + node.state + " with a cost of " + str(node.path_cost) + ".")
        path = node.path()
        directions = ""
        for n in path:
            directions = directions + " > " + n.state
        print("The path is the following:" + directions)


In [33]:
sbb = SBB()
sbb.importData('linie-mit-betriebspunkten.json')

start = 'Hitzkirch'
goal = 'Bellinzona'
sbb_map = UndirectedGraph(sbb.createMap())
problem = GraphProblem(start, goal, sbb_map)

successfully imported 2787 hubs
successfully imported 401 train lines


In [34]:
a_star_solution = a_star_algorithm(problem, sbb)

Reached goal destination: Bellinzona
Nodes visited: 425
Max nodes stored: 42524


In [35]:
print_info_about_search(a_star_solution)

The search algorithm reached Bellinzona with a cost of 190.543.
The path is the following: > Hitzkirch > Gelfingen > Baldegg > Baldegg_Kloster > Hochdorf > Hochdorf_Schonau > Ballwil > Eschenbach > Waldibrucke > Hubeli_LU > Emmenbrucke_Gersag > Emmenbrucke > Fluhmuhle_Abzw > Gutsch_Abzw > Luzern_Verkehrshaus > Meggen_Zentrum > Meggen > Merlischachen > Kussnacht_am_Rigi > Immensee_West_Abzw > Immensee > Brunnmatt_Spw > Arth-Goldau > Arth-Goldau_Ost_Abzw > Steinen > Schwyz > Brunnen > Brunnen_Sud_Abzw > Sisikon_Nord > Sisikon > Gruonbach_Spw > Fluelen > Altdorf > Rynacht_Nord > Rynacht_Abzw > Erstfeld_Nord_Abzw > Erstfeld > Amsteg-Silenen > Zgraggen_Spw > Gurtnellen > Pfaffensprung_Spw > Wassen > Eggwald_Spw > Goschenen > Gotthard_Nord_Spw > Gottardo_Sud_c_bin > Airolo > Sordo_c_bin > Ambri-Piotta > Rodi-Fiesso > Pardorea_c_bin > Faido > Chiggiogna_c_bin > Lavorgo > Pianotondo_c_bin > Giornico_c_bin > Giornico > Bodio_TI > Pollegio_Nord_dira > Biasca > Giustizia_dira > Osogna-Cresciano >

In [36]:
ida_star_solution = ida_star_algorithm(problem, sbb)

Reached goal destination: Bellinzona
Nodes visited: 425
Max nodes stored: 42524


In [37]:
print_info_about_search(ida_star_solution)

The search algorithm reached Bellinzona with a cost of 190.543.
The path is the following: > Hitzkirch > Gelfingen > Baldegg > Baldegg_Kloster > Hochdorf > Hochdorf_Schonau > Ballwil > Eschenbach > Waldibrucke > Hubeli_LU > Emmenbrucke_Gersag > Emmenbrucke > Fluhmuhle_Abzw > Gutsch_Abzw > Luzern_Verkehrshaus > Meggen_Zentrum > Meggen > Merlischachen > Kussnacht_am_Rigi > Immensee_West_Abzw > Immensee > Brunnmatt_Spw > Arth-Goldau > Arth-Goldau_Ost_Abzw > Steinen > Schwyz > Brunnen > Brunnen_Sud_Abzw > Sisikon_Nord > Sisikon > Gruonbach_Spw > Fluelen > Altdorf > Rynacht_Nord > Rynacht_Abzw > Erstfeld_Nord_Abzw > Erstfeld > Amsteg-Silenen > Zgraggen_Spw > Gurtnellen > Pfaffensprung_Spw > Wassen > Eggwald_Spw > Goschenen > Gotthard_Nord_Spw > Gottardo_Sud_c_bin > Airolo > Sordo_c_bin > Ambri-Piotta > Rodi-Fiesso > Pardorea_c_bin > Faido > Chiggiogna_c_bin > Lavorgo > Pianotondo_c_bin > Giornico_c_bin > Giornico > Bodio_TI > Pollegio_Nord_dira > Biasca > Giustizia_dira > Osogna-Cresciano >

Hints: you can use the heap library heapq for your frontier:

`from heapq import heappush, heappop`

The following line will add the node `f` to the frontier with priority `f`:

`heappush(frontier, (f, node))`

To get the first node, use: `node = heappop(frontier)[1]`

The aerial distance between a node and the goal can be computed with the following function:

`sbb.get_distance_between(node.state, problem.goal)`
        

How do theses informed search algorithms perform on our problem? Create the following overview table for the example problem.


| Algorithm | start   | goal | cost | number of nodes visited | maximal stored nodes | complete | optimal |
|------|------|-----|-----|-----|-----|-----|-----|
| A*|Hitzkirch|Bellinzona|190.543|425|42524|yes|no
| IDA*|Hitzkirch|Bellinzona|190.543|425|42524|yes|yes
